<a href="https://colab.research.google.com/github/AlienX77-cmd/Python-for-Business-and-Finance/blob/main/Exercise_05.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Exercises (Part 1)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 5: While Loops, Conditional Statements, IRR and YTM

The ABC Company has a mature project with declining revenues and increasing costs. Net Cashflows are projected as follows:

- 500 in year 1,
- 350 in year 2, 
- 200 in year 3,
- 50 in year 4,
- -100 in year 5,
- -200 in year 6

1. Separate __positive__ and __negative__ cashflows. __Create two new lists__!

In [ ]:
#run the cell
cf = [500, 350, 200, 50, -100, -200]

In [ ]:
cf_pos =[]
cf_neg =[]

for i in cf: #complete the code!
    if (i > 0): cf_pos.append(i)
    else: cf_neg.append(i)   

print(cf_pos, cf_neg)    

[500, 350, 200, 50] [-100, -200]


2. Get the __first unprofitable year__. Your Code should return the statement __"Year ... is the first unprofitable Year!"__ 

In [ ]:
for i in range(len(cf)):
  if(cf[i] < 0):
    print("Year {} is the first unprofitable Year!".format(i+1))
    break

Year 5 is the first unprofitable Year!


3. Get the __last profitable year__. Your Code should return the statement __"Year ... is the last profitable Year!"__ 

In [ ]:
for i in range(len(cf)):
  if(cf[i] > 0): pass
  else:
    print("Year {} is the last profitable Year!".format(i))
    break

Year 4 is the last profitable Year!


The __modulo operator %__ returns the __remainder__ of the division between two numbers. For instance, 5 % 3 returns 2. 

In [ ]:
# run the cell
5 % 3

2

In [ ]:
# run the cell
14 % 5

4

If a is divisible by b __without remainder__, a % b __returns 0__. 

In [ ]:
# run the cell
a = 9
b = 3
a % b

0

4. Return the __first number n__ that is __divisible by 362 and 216__ (without remainder).<br>
Your code shall return the following statement: __"... is the first number divisible by 362 and 216!"__

In [ ]:
a = 362
b = 216

n = 1 
while True: #complete the code!
    if (n % 362 == 0):
      if (n % 216 == 0): 
        print("{} is the first number divisible by 362 and 216".format(n))
        break
    n+=1

39096 is the first number divisible by 362 and 216


(result: 39096)

The ABC Company evaluates to buy an additional machine that will increase profits/cashflows by
- 50 USD in year 1,
- 75 USD in year 2 ,
- 100 USD in year 3, 
- 150 USD in year 4,
- 125 USD in year 5. (each cf at period´s end)

The machine costs 450 USD (Investment in to). 

5. Calculate the __Project´s IRR__ and evaluate whether ABC should pursue the project. ABCs required rate of return is 6% p.a.

In [15]:
cf = [-450, 50, 75, 100, 150, 125]

In [16]:
guess = 0.06
step = 0.000001
target_npv = 0
tol = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tol:
        if diff > 0: guess += step
        elif diff < 0: guess -= step
    else:
        break
        
print(NPV, round(guess*100, 2)) 

-0.0002819830025373449 3.12


(result: 3.12%)

5. Would your __decision change__ if you can increse year 5 cashflows from __125 to 175__? 

In [17]:
cf[5] = 175
cf

[-450, 50, 75, 100, 150, 175]

In [18]:
guess = 0.06
step = 0.000001
target_npv = 0
tol = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tol:
        if diff > 0: guess += step
        elif diff < 0: guess -= step
    else:
        break
        
print(NPV, round(guess*100, 2)) 

-0.0003708748203337109 5.83


(result: 5.83%)

6. The ABC Government issued a 30Y Government Bond one year ago with a Coupon Rate of 5.5% (annual payments in arrears). Today´s Bond Price is 97.2 (per 100 par Value). Calculate the Bond´s __current YTM__.

In [27]:
cf = [5.5] * 29
print(cf, sep=", ")

[5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5]


In [28]:
cf[-1] = cf[-1] + 100
print(cf, sep=", ")

[5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 105.5]


In [29]:
cf.insert(0, -97.2)

In [30]:
print(cf, sep=", ")

[-97.2, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 5.5, 105.5]


In [31]:
#Caculate IRR (IRR = YTM in this case (normally))
guess = 0.06
step = 0.000001
target_npv = 0
tol = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tol:
        if diff > 0: guess += step
        elif diff < 0: guess -= step
    else:
        break
        
print(NPV, round(guess*100, 2)) 

-0.00023920511386066323 5.7


(result: 5.70%)

In [32]:
print("Make sense as normally IRR (YTM) is higher than coupon rate and in this case, we initially invested 97.2 $ in the first year and recieve 5 $ each year and recieve 105 $ in the last year (more than 97.2 $)")

Make sense as normally IRR (YTM) is higher than coupon rate and in this case, we initially invested 97.2 $ in the first year and recieve 5 $ each year and recieve 105 $ in the last year (more than 97.2 $)


# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

The ABC Company has a mature project with declining revenues and increasing costs. Net Cashflows are projected as follows:

- 500 in year 1,
- 350 in year 2, 
- 200 in year 3,
- 50 in year 4,
- -100 in year 5,
- -200 in year 6

1. Separate __positive__ and __negative__ cashflows. __Create two new lists__!

In [ ]:
#run the cell
cf = [500, 350, 200, 50, -100, -200]

In [ ]:
cf_pos =[]
cf_neg =[]

for i in cf:
    if i > 0:
        cf_pos.append(i)
    elif i < 0:
        cf_neg.append(i)
print(cf_pos, cf_neg)

2. Get the __first unprofitable year__. Your Code should return the statement __"Year ... is the first unprofitable Year!"__ 

In [ ]:
n = 1
for i in cf:
    if i < 0:
        print("Year {} is the first unprofitable year!".format(n))
        break
    n += 1

3. Get the __last profitable year__. Your Code should return the statement __"Year ... is the last profitable Year!"__ 

In [ ]:
n = 1
for i in cf:
    if i < 0:
        print("Year {} is the last profitable year!".format(n - 1))
        break
    n += 1

The __modulo operator %__ returns the __remainder__ of the division between two numbers. For instance, 5 % 3 returns 2. 

In [ ]:
# run the cell
5 % 3

In [ ]:
# run the cell
14 % 5

If a is divisible by b __without remainder__, a % b __returns 0__. 

In [ ]:
# run the cell
a = 9
b = 3
a % b

4. Return the __first number n__ that is __divisible by 362 and 216__ (without remainder).<br>
Your code shall return the following statement: __"... is the first number divisible by 362 and 216!"__

In [ ]:
a = 362
b = 216

n = 1
while True:
    if (n%a == 0) and (n%b == 0):
        print("{} is the first number divisible by {} and {}!".format(n, a, b))
        break
    n+=1

(result: 39096)

The ABC Company evaluates to buy an additional machine that will increase profits/cashflows by
- 50 USD in year 1,
- 75 USD in year 2 ,
- 100 USD in year 3, 
- 150 USD in year 4,
- 125 USD in year 5. (each cf at period´s end)

The machine costs 450 USD (Investment in to). 

5. Calculate the __Project´s IRR__ and evaluate whether ABC should pursue the project. ABCs required rate of return is 6% p.a.

In [ ]:
cf = [-450, 50, 75, 100, 150, 125]

In [ ]:
guess = 0.06
step = 0.000001
target_npv = 0
tolerance = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tolerance:
        if diff < 0:
            guess -= step
        elif diff > 0:
            guess += step
    else:
        break
        
print(NPV, guess) 

(result: 3.12% -> reject)

5. Would your __decision change__ if you can increse year 5 cashflows from __125 to 175__? 

In [ ]:
cf[-1] = 175

In [ ]:
guess = 0.06
step = 0.0000001
target_npv = 0
tolerance = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tolerance:
        if diff < 0:
            guess -= step
        elif diff > 0:
            guess += step
    else:
        break
        
print(NPV, guess) 

(result: 5.83% -> still reject)

6. The ABC Government issued a 30Y Government Bond one year ago with a Coupon Rate of 5.5% (annual payments in arrears). Today´s Bond Price is 97.2 (per 100 par Value). Calculate the Bond´s __current YTM__.

In [ ]:
cf = [5.5] * 29

In [ ]:
cf[-1] = cf[-1] + 100

In [ ]:
cf.insert(0, -97.2)

In [ ]:
cf

In [ ]:
guess = 0.06
step = 0.0000001
target_npv = 0
tolerance = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tolerance:
        if diff < 0:
            guess -= step
        elif diff > 0:
            guess += step
    else:
        break
        
print(NPV, guess) 

(result: 5.70%)